private/pkg/manifest/storage.go

```diff
@@ -0,0 +1,190 @@
+func (o *manifestBucketObject) Close() error               { return o.file.Close() }
+
+// NewFromBucket creates a manifest from a storage bucket, with all its digests
+// in DigestTypeShake256.
+func NewFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := New()
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) (retErr error) {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		defer func() { retErr = multierr.Append(retErr, obj.Close()) }()
+		digest, err := digester.Digest(obj)
+		if err != nil {
+			return err
+		}
+		if err := m.AddEntry(path, *digest); err != nil {
+			return err
+		}
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+	return m, nil
+}
+
+type bucketOptions struct {
+	allManifestBlobs bool
+	noExtraBlobs     bool
+}
+
+// BucketOption are options passed when creating a new manifest bucket.
+type BucketOption func(*bucketOptions)
+
+// BucketWithAllManifestBlobsValidation validates that all the passed digests in
+// the manifest have a corresponding blob in the blob set. If this option is not
+// passed, then buckets with partial/incomplete blobs is allowed.
+func BucketWithAllManifestBlobsValidation() BucketOption {
+	return func(opts *bucketOptions) {
+		opts.allManifestBlobs = true
+	}
+}
+
+// BucketWithNoExtraBlobsValidation validates that the passed blob set has no
+// additional blobs beyond the ones in the manifest.
+func BucketWithNoExtraBlobsValidation() BucketOption {
+	return func(opts *bucketOptions) {
+		opts.noExtraBlobs = true
+	}
+}
+
+// NewBucket takes a manifest and a blob set and builds a readable storage
+// bucket that contains the files in the manifest.
+func NewBucket(m Manifest, blobs BlobSet, opts ...BucketOption) (storage.ReadBucket, error) {
+	var config bucketOptions
+	for _, option := range opts {
+		option(&config)
+	}
+	if config.allManifestBlobs {
+		for _, path := range m.Paths() {
+			pathDigest, ok := m.DigestFor(path)
+			if !ok {
+				// we're iterating manifest paths, this should never happen.
+				return nil, fmt.Errorf("path %q not present in manifest", path)
+			}
+			if _, ok := blobs.digestToBlob[pathDigest.String()]; !ok {
+				return nil, fmt.Errorf("manifest path %q with digest %q has no associated blob", path, pathDigest.String())
+			}
+		}
+	}
+	if config.noExtraBlobs {
+		for digestStr := range blobs.digestToBlob {
+			if _, ok := m.digestToPaths[digestStr]; !ok {
+				return nil, fmt.Errorf("blob with digest %q is not present in the manifest", digestStr)
+			}
+		}
+	}
+	return &manifestBucket{
+		manifest: m,
+		blobs:    blobs,
+	}, nil
+}
+
+func (m *manifestBucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	digest, ok := m.manifest.pathToDigest[path]
```

Comment:
IMHO, I'd use the equivalent DigestFor. Less private member access, even if we're in the same module.

```suggestion
	digest, ok := m.manifest.DigestFor(path)
```

Reply:
Did `manifestBucket.blobFor(path)` suggestion instead.